File: week4_C/q1.c
#include <stdio.h>
#include <stdlib.h>
struct Node { int data; struct Node* prev; struct Node* next; };
struct Node* head = NULL;
void insertAtBeginning(int data) {
    struct Node* newNode = malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->prev = NULL;
    newNode->next = head;
    if (head != NULL) head->prev = newNode;
    head = newNode;
}
void display() {
    struct Node* temp = head;
    while (temp) { printf("%d ", temp->data); temp = temp->next; }
}
int main() {
    insertAtBeginning(3);
    insertAtBeginning(2);
    insertAtBeginning(1);
    display();
}

File: week4/q1.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None
head = None
def insert_at_beginning(data):
    global head
    new_node = Node(data)
    new_node.next = head
    if head:
        head.prev = new_node
    head = new_node
def display():
    temp = head
    while temp:
        print(temp.data, end=" ")
        temp = temp.next
insert_at_beginning(3)
insert_at_beginning(2)
insert_at_beginning(1)
display()